edgelet/hsm-sys/azure-iot-hsm-c/tests/edge_openssl_int/edge_openssl_int.c

```diff
@@ -1038,4 +1042,107 @@ BEGIN_TEST_SUITE(edge_openssl_int_tests)
+
+        // assert
+        ASSERT_ARE_EQUAL(int, 0, result, "Line:" TOSTRING(__LINE__));
+        // if this assertion fails it implies that the call to generate_rand_buffer
+        // never updated the buffer and yet returned a success OR
+        // the statistically improbable event occured that the random bytes returned
+        // exactly what the unexpected_buffer of size N was setup with
+        // P(test failure) = P(0xF1) * P(0xF1) * ... * P(0xF1) = ((1/256) ^ N) == very small
+        ASSERT_ARE_NOT_EQUAL(int, 0, memcmp(unexpected_buffer, output_buffer, buffer_sz), "Line:" TOSTRING(__LINE__));
+
+        //cleanup
+    }
+
+    TEST_FUNCTION(test_rand_large_buf)
+    {
+        // arrange
+        unsigned char unexpected_buffer[TEST_RAND_SIZE_BYTES_LARGE];
+        unsigned char output_buffer[TEST_RAND_SIZE_BYTES_LARGE];
+        size_t buffer_sz = sizeof(unexpected_buffer);
+
+        memset(unexpected_buffer, 0xF1, buffer_sz);
+        memset(output_buffer, 0xF1, buffer_sz);
+
+        // act
+        int result = generate_rand_buffer(output_buffer, buffer_sz);
+
+        // assert
+        ASSERT_ARE_EQUAL(int, 0, result, "Line:" TOSTRING(__LINE__));
+        // if this assertion fails it implies that the call to generate_rand_buffer
+        // never updated the buffer and yet returned a success OR
+        // the statistically improbable event occured that the random bytes returned
+        // exactly what the unexpected_buffer of size N was setup with
+        // P(test failure) = P(0xF1) * P(0xF1) * ... * P(0xF1) = ((1/256) ^ N) == very small
+        ASSERT_ARE_NOT_EQUAL(int, 0, memcmp(unexpected_buffer, output_buffer, buffer_sz), "Line:" TOSTRING(__LINE__));
+
+        //cleanup
+    }
+
+    TEST_FUNCTION(test_rand_multiple_calls)
+    {
+        // arrange
+        int result_1, result_2;
+        unsigned char output_buffer_1[TEST_RAND_SIZE_BYTES_LARGE];
+        unsigned char output_buffer_2[TEST_RAND_SIZE_BYTES_LARGE];
+        size_t buffer_sz = sizeof(output_buffer_1);
+
+        memset(output_buffer_1, 0xF1, buffer_sz);
+        memset(output_buffer_2, 0xF1, buffer_sz);
+
+        // act
+        result_1 = generate_rand_buffer(output_buffer_1, buffer_sz);
+        result_2 = generate_rand_buffer(output_buffer_2, buffer_sz);
+
+        // assert
+        ASSERT_ARE_EQUAL(int, 0, result_1, "Line:" TOSTRING(__LINE__));
+        ASSERT_ARE_EQUAL(int, 0, result_2, "Line:" TOSTRING(__LINE__));
+        // if this assertion fails it implies that the call to generate_rand_buffer
+        // never updated the buffer and yet returned a success OR
+        // the statistically improbable event occured that the random bytes returned
+        // exactly what the output_buffer_2 of size N was setup with
+        // P(test failure) = P(0xF1) * P(0xF1) * ... * P(0xF1) = ((1/256) ^ N) == very small
+        ASSERT_ARE_NOT_EQUAL(int, 0, memcmp(output_buffer_1, output_buffer_2, buffer_sz), "Line:" TOSTRING(__LINE__));
```

Comment:
>ASSERT_ARE_NOT_EQUAL [](start = 8, length = 20)

This is kind of a minimum bar test for a random function. 
 I was thinking it might be good to have a test that raises this a bit.  

That means a statistical approach to the test, should we implement one or more of these?
[https://csrc.nist.gov/Projects/Random-Bit-Generation/Documentation-and-Software/Guide-to-the-Statistical-Tests](https://csrc.nist.gov/Projects/Random-Bit-Generation/Documentation-and-Software/Guide-to-the-Statistical-Tests)

Ore should we just have trust in openssl's library?

Reply:
The idea here is that we should trust openssl and a HW vendor's implementation of /dev/random or /dev/urandom. Remember we port on OpenSSL. The implementation prior to this used the C runtime's implementation of rand() which I am not sure we could reliably take a dependency on to get a statistically good random number.